internal/identity/providers.go

```diff
@@ -192,6 +195,58 @@ func (p *Provider) IdentityFromToken(ctx context.Context, t *oauth2.Token) (*oid
 
 // Revoke enables a user to revoke her token. If the identity provider supports revocation
 // the endpoint is available, otherwise an error is thrown.
+//
+// We will attempt to get the identity provider's possible revocation endpoint from
+// their /.well-known/openid-configuration.
+// we use the revocation_endpoint or end_session_endpoint to implement
+// the generic revocation of a user's access token.
+//
+// reference:
+// Google : https://accounts.google.com/.well-known/openid-configuration
+// Azure: https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration
 func (p *Provider) Revoke(ctx context.Context, token *oauth2.Token) error {
-	return ErrRevokeNotImplemented
+	if p.RevocationURL == "" {
+		return ErrRevokeNotImplemented
+	}
+	params := url.Values{}
+	params.Add("token", token.AccessToken)
+	err := httputil.Client(ctx, http.MethodPost, p.RevocationURL, version.UserAgent(), nil, params, nil)
+	if err != nil && err != httputil.ErrTokenRevoked {
+		return err
+	}
+
+	return nil
+}
+
+// UnmarshalJSON method is needed so that provider will be able to
+// satisfy the Unmarshaler interface
+// https://pkg.go.dev/gopkg.in/square/go-jose.v2/json#Unmarshaler
+//
+// For the the go-oidc package to be able to apply the Claims method which tries to unmarshal
+// the response body from the provider's /.well-known/openid-configuration
+// on the supplied interface, and in our case, it will be the Provider.
+// An extra step is made to create this implementation because the endpoint
+// to revoke a user's access token can be revocation_endpoint or end_session_endpoint.
+// We want to be able to have only one field and not repeat any of the two possible
+// endpoints on the Provider field.
+func (p *Provider) UnmarshalJSON(b []byte) error {
+	type Alias Provider
+	t := &struct {
+		*Alias
+		EndSessionEndpoint string `json:"end_session_endpoint,omitempty"`
+		RevocationEndpoint string `json:"revocation_endpoint,omitempty"`
+	}{
+		Alias: (*Alias)(p),
+	}
+	if err := json.Unmarshal(b, &t); err != nil {
+		return err
+	}
+	if t.EndSessionEndpoint != "" {
+		t.RevocationURL = t.EndSessionEndpoint
+	}
+	if t.RevocationEndpoint != "" {
+		t.RevocationURL = t.RevocationEndpoint
+	}
+	*p = *(*Provider)(t.Alias)
```

Comment:
This is a really nice solution!

Is this particular line necessary? I think it updates the original fields via the pointer.

Also could we add a unit test to make sure `end_session_endpoint` and `revocation_endpoint` get handled properly? Seems like this might be an edge case that could easily be dropped accidentally in the future.